### PR TITLE
feat(core): add validation and builder pattern to LintConfig (#352)

### LINKED ISSUE #352: Add validation and builder pattern to LintConfig
## Context
From code audit (architecture review). `LintConfig` has public fields with no invariant checking.

## Current State
- 11 public fields with no validation
- Runtime state (`root_dir`, `import_cache`) mixed with serialized config
- No builder pattern - callers construct manually
- Setting invalid values doesn't error until runtime

## Proposal
1. Use builder pattern: `LintConfigBuilder` with validation on `build()`
2. Make fields private with accessor methods
3. Separate `Config` (serializable) from `RuntimeContext` (runtime state)
4. Validate on load, not on access

## Location
`crates/agnix-core/src/config.rs:311-415`

## Effort
High - significant refactoring but improves safety and API clarity

### PR DESCRIPTION
**Pull request by avifenesh**
## Summary

- Add `LintConfigBuilder` with fluent `&mut Self` API following the existing `ValidatorRegistryBuilder` pattern
- Make all 11 serializable fields on `LintConfig` private with getter/setter methods
- Move `root_dir` and `import_cache` into `RuntimeContext`, completing config/runtime separation
- Add `ConfigError` enum for build-time validation (invalid globs, path traversal, unknown tools/rules)
- Validate on construction via `build()`, not on access
- Export `LintConfigBuilder` and `ConfigError` from `agnix-core` public API

### New Types
- `LintConfigBuilder` - Builder with `build() -> Result<LintConfig, ConfigError>` and `build_unchecked()` for tests
- `ConfigError` - `InvalidGlobPattern`, `PathTraversal`, `ValidationFailed`

### API Migration
| Before | After |
|--------|-------|
| `config.severity` | `config.severity()` |
| `config.rules.field` | `config.rules().field` |
| `config.target = X` | `config.set_target(X)` |
| `config.rules.field = Y` | `config.rules_mut().field = Y` |
| `LintConfig::default()` | `LintConfig::builder().build()?` |

### Validation in build()
- Glob pattern syntax (exclude + all FilesConfig lists)
- Path traversal detection (`..`, `/../`, `/..`)
- Tool name validation against known tools
- Rule ID prefix validation
- Deprecated field warnings promoted to errors

## Test Plan
- 49 new tests (34 builder unit + 15 edge cases for review findings)
- All 2,534 existing tests pass
- Cross-crate integration tests verify builder API from outside agnix-core
- Serde roundtrip tests confirm private fields still deserialize correctly

## Files Changed
35 files across all crates (config.rs is primary, rest are field access migrations)

Closes #352

### CODE REVIEW

File: crates/agnix-core/src/config.rs [hunk 1]
```diff
@@ -632,7 +675,292 @@ pub enum TargetTool {
...
+        }
+        if !self.disabled_validators.is_empty() {
+            let mut seen: std::collections::HashSet<String> =
+                rules.disabled_validators.iter().cloned().collect();
+            for name in self.disabled_validators.drain(..) {
+                if seen.insert(name.clone()) {
+                    rules.disabled_validators.push(name);
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/config.rs:920]: Applied, thanks\! Used \ + \ with HashSet as suggested.

[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/config.rs:920]: Applied, thanks\! Used append + retain with HashSet as suggested.

File: CLAUDE.md [hunk 2]
```diff
@@ -97,6 +97,23 @@ impl ValidatorRegistry {
...
+impl LintConfigBuilder {
+    pub fn new() -> Self;
+    pub fn severity(&mut self, s: SeverityLevel) -> &mut Self;
+    pub fn target(&mut self, t: TargetTool) -> &mut Self;
+    pub fn tools(&mut self, t: Vec<String>) -> &mut Self;
+    pub fn exclude(&mut self, e: Vec<String>) -> &mut Self;
+    pub fn disable_rule(&mut self, id: impl Into<String>) -> &mut Self;
+    pub fn disable_validator(&mut self, name: impl Into<String>) -> &mut Self;
+    pub fn build(&mut self) -> Result<LintConfig, ConfigError>;
+    pub fn build_unchecked(&mut self) -> LintConfig;
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply CLAUDE.md:113]: Fixed. Updated to show LintConfig::builder() factory method and removed the private new() from the docs.

File: AGENTS.md [hunk 3]
```diff
@@ -97,6 +97,23 @@ impl ValidatorRegistry {
...
+impl LintConfigBuilder {
+    pub fn new() -> Self;
+    pub fn severity(&mut self, s: SeverityLevel) -> &mut Self;
+    pub fn target(&mut self, t: TargetTool) -> &mut Self;
+    pub fn tools(&mut self, t: Vec<String>) -> &mut Self;
+    pub fn exclude(&mut self, e: Vec<String>) -> &mut Self;
+    pub fn disable_rule(&mut self, id: impl Into<String>) -> &mut Self;
+    pub fn disable_validator(&mut self, name: impl Into<String>) -> &mut Self;
+    pub fn build(&mut self) -> Result<LintConfig, ConfigError>;
+    pub fn build_unchecked(&mut self) -> LintConfig;
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply AGENTS.md:113]: Fixed. AGENTS.md kept identical to CLAUDE.md.

File: crates/agnix-lsp/src/vscode_config.rs [hunk 4]
```diff
@@ -558,11 +558,11 @@ mod tests {
         vscode_config.merge_into_lint_config(&mut lint_config);
 
         assert_eq!(
-            lint_config.tool_versions.claude_code,
+            lint_config.tool_versions_mut().claude_code,
             Some("1.0.0".to_string())
         );
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-lsp/src/vscode_config.rs:563]: Fixed. Changed to tool_versions() immutable accessor.

File: tests/cross_crate_integration.rs [hunk 5]
```diff
@@ -450,3 +450,100 @@ fn validator_name_accessible_from_outside_crate() {
...
+#[test]
+fn builder_with_root_dir_used_by_validate_project() {
+    let dir = tempfile::tempdir().unwrap();
+    let config = agnix_core::LintConfig::builder()
+        .root_dir(dir.path().to_path_buf())
+        .build()
+        .unwrap();
+
+    let result = agnix_core::validate_project(dir.path(), &config);
+    assert!(result.is_ok());
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply tests/cross_crate_integration.rs]: Renamed to builder_config_works_with_validate_project to avoid implying root_dir carries through.

File: crates/agnix-lsp/src/backend.rs [hunk 6]
```diff
@@ -591,7 +591,7 @@ impl LanguageServer for Backend {
                                 crate::locale::init_from_config(config_locale);
                             }
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-lsp/src/backend.rs:592]: Good catch\! This was a real bug that slipped through because agnix-lsp was excluded from compilation on Windows. Fixed to use loaded_config.locale() accessor.

File: crates/agnix-core/src/config.rs [hunk 7]
```diff
@@ -124,6 +124,50 @@ pub struct FilesConfig {
...
+            }
+            ConfigError::PathTraversal { pattern } => {
+                write!(f, "path traversal in pattern '{}'", pattern)
+            }
+            ConfigError::ValidationFailed(warnings) => {
+                write!(
+                    f,
+                    "configuration validation failed with {} warning(s)",
+                    warnings.len()
+                )
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/config.rs]: Fixed. Now includes the first warning message in the Display output for better diagnostics.

File: crates/agnix-core/src/config.rs [hunk 8]
```diff
@@ -632,7 +682,286 @@ pub enum TargetTool {
...
+        Ok(config)
+    }
+
+    /// Build the `LintConfig` without running any validation.
+    ///
+    /// This is primarily intended for tests that need to construct configs
+    /// with intentionally invalid data.
+    pub fn build_unchecked(&mut self) -> LintConfig {
+        self.build_inner()
+    }
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/config.rs:900]: Added doc note that build()/build_unchecked() drain state, making the builder single-use. This matches the ValidatorRegistryBuilder pattern.

File: crates/agnix-core/src/pipeline.rs [hunk 9]
```diff
@@ -172,23 +172,24 @@ fn resolve_with_compiled(
...
+    let files = config.files_config();
+    if files.include_as_memory.is_empty()
+        && files.include_as_generic.is_empty()
+        && files.exclude.is_empty()
     {
         return detect_file_type(path);
     }
 
     // Compile patterns on-demand for single-file validation.
     // Invalid patterns are silently skipped (validated at config load time).
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/pipeline.rs]: Fixed. Updated comment to accurately reflect behavior: patterns are silently skipped here, pointing users to LintConfigBuilder::build() or LintConfig::validate() for strict validation.
